### PR TITLE
More tests

### PR DESCRIPTION
**Pull request by chronitis**
More tests:
- Check `inspect_request` returns something
- Test execution with `silent=True` is silent
- Test several aspects of history: `tail`, `range` and `search` queries, and some of their options (`output`, `n`, `unique`). There are more combinations possible here, but I'm not actually sure what the correct semantics should be in some cases.
- Test for `file_extension` in `kernel_info_reply`
- Test that an error is reported in both `execute_reply` and an `error` message

In addition, some existing tests have had the `unittest.subTest()` context manager inserted where multiple aspects or code inputs were being tested to make the cause of errors clearer.

### CODE REVIEW

File: jupyter_kernel_test/__init__.py [hunk 1]
```diff
@@ -96,12 +105,13 @@ def test_completion(self):
             raise SkipTest
 
         for sample in self.completion_samples:
-            msg_id = self.kc.complete(sample['text'])
-            reply = self.kc.get_shell_msg()
-            validate_message(reply, 'complete_reply', msg_id)
-            if 'matches' in sample:
-                self.assertEqual(set(reply['content']['matches']),
-                                 set(sample['matches']))
+            with self.subTest(text=sample['text']):
```
Comment (takluyver) [inline jupyter_kernel_test/__init__.py:123]: Neat, I didn't know about this.

File: jupyter_kernel_test/__init__.py [hunk 2]
```diff
@@ -188,11 +204,111 @@ def test_display_data(self):
...
+        msg_id = self.kc.history(**histargs)
+
+        reply = self.kc.get_shell_msg(timeout=timeout)
+        validate_message(reply, 'history_reply', msg_id)
+
+        return reply
+
+    def test_history(self):
+        if not self.code_execute_result:
+            raise SkipTest
```
Comment (takluyver) [inline jupyter_kernel_test/__init__.py:252]: We certainly also need at least a separate flag for the history tests, and possibly separate switches for different types of history access. The notebook never uses history requests, so many kernels don't implement them. And `jupyter console` and the qtconsole only use `tail` requests to get the history you see when you press the up arrow.

I think we just exposed all the history operations IPython supports, rather than building up what frontends actually need. We should probably think more about it at some point, but it works, so we've never really looked at it much.

File: jupyter_kernel_test/__init__.py [hunk 3]
```diff
@@ -188,11 +204,111 @@ def test_display_data(self):
...
+                self.assertEqual(reply['content']['status'], 'ok')
+
+                self.assertGreaterEqual(len(output_msgs), 1)
+                self.assertEqual(output_msgs[0]['msg_type'], 'display_data')
+                self.assertIn(sample['mime'], output_msgs[0]['content']['data'])
+
+    def test_silent(self):
+        self.flush_channels()
+
+        reply, output_msgs = self.execute_helper("", silent=True)
```
Comment (takluyver) [inline jupyter_kernel_test/__init__.py]: Ditto for silent, there may well be kernels that don't implement it. Also, we should probably reuse the hello world sample, because:
- You might get bizarre behaviour when executing an empty string. I think our frontends might even short circuit that. We should probably decide and document it.
- If you don't hit corner cases, it's not producing any output anyway, so you're not really testing the silent flag.

So far, the design of jkt is that kernels opt in to each test by specifying an attribute on their subclass. That's up for discussion, but I'm inclined to keep the principle - it means we can use the latest version of jkt to test IRkernel, for instance, without worrying that new tests will suddenly appear and start breaking things.

File: jupyter_kernel_test/__init__.py (hunk 1 above)
Comment (chronitis) [inline jupyter_kernel_test/__init__.py:123]: Neither did I, until today (new in 3.4).

File: jupyter_kernel_test/__init__.py (hunk 3 above)
Comment (chronitis) [inline jupyter_kernel_test/__init__.py]: The messaging protocol documentation implies this should be a supported operation

> In order to obtain the current execution counter for the purposes of displaying input prompts, frontends may make an execution request with an empty code string and `silent=True`.

but I take the point that this doesn't actually very well test the semantics, and that it would be better only to have tests with corresponding attributes. I'll see if I can think of a better approach for this one or drop it.

File: jupyter_kernel_test/__init__.py (hunk 2 above)
Comment (chronitis) [inline jupyter_kernel_test/__init__.py:252]: I originally had separate tests for the different history modes, and bundled them into a single function with subtests to reduce code duplication. Possibly separating them and adding multiple flags would be better here? I assume that given that history is (for now) documented in the messaging protocol you're happy to carry tests for it.

Comment (takluyver) [inline jupyter_kernel_test/__init__.py:252]: Yep, happy to have tests for it, but I don't want to run them unless the subclass explicitly indicates that we should, as I think most kernels don't implement the history messages.

**Comment by chronitis** [comment]
- History operations are now controlled by listing access types to test in `supported_history_operations` (defaulting to skipping everything)
- The silent execution test is removed for now since I couldn't think of a good way of doing it
- Additional trivial tests: `stream name=stderr` and `clear_output` messages (and corresponding code attributes)

File: README.rst [hunk 4]
```diff
@@ -20,11 +20,18 @@ Use it like this:
         # language_info.name in a kernel_info_reply should match this
         language_name = "dg"
 
+        # language_info.file_extension, should match kernel_info_reply
+        # (and start with a dot)
+        file_extension = ".py"
```
Comment (takluyver) [inline README.rst:25]: The example in the README was originally for [dogelang](https://pyos.github.io/dg/), to highlight that it's not just for Python. I think it's now become a bit of a confusing hodgepodge of that and Python, and dogelang itself has probably changed since I wrote the example (since it's not exactly a very serious thing).

That can be cleaned up in a separate PR, though.

File: jupyter_kernel_test/__init__.py [hunk 5]
```diff
@@ -188,11 +219,124 @@ def test_display_data(self):
...
+
+    def test_inspect(self):
+        if not self.code_inspect_sample:
+            raise SkipTest
+
+        for i in (0, 1):
+            with self.subTest(detail_level=i):
+                self.flush_channels()
+                msg_id = self.kc.inspect(self.code_inspect_sample,
+                                         detail_level=i)
```
Comment (takluyver) [inline jupyter_kernel_test/__init__.py]: detail_level is another thing that I don't know how far different kernels implement. Again, we copied part of the IPython infrastructure into the message protocol, and never really thought about whether it made sense. I'm not even sure if any of our frontends use `detail_level=1`.

Let's just test it with the default level for now. I'd rather start with fewer tests and add them as we determine they're useful than add more and decide to remove things.

**Comment by chronitis** [comment]
Inspection now only tested at `detail=0`.

In an unrelated note, a future version of the messaging spec might benefit from some RFC2119-style _MUST_, _SHOULD_, _MAY_-type formalisation on what is required for a conforming kernel implementation and what is optional, both in terms of message contents and kernel behaviour. Testing several of the kernels I packaged with this shows a number of deviations from the spec as written (while still successfully interacting with the notebook), and I'm unsure if these should be regarded as bugs in the kernels or in the spec.

**Comment by takluyver** [comment]
Thanks!

Yes, the spec is definitely somewhat ad-hoc at present. I'd say that only `kernel_info` and `execute` messages are required to be implemented, and several of the fields in `execute` can be ignored without much problem. I think it would definitely make sense to formalise that, though.
